File: cpp/3904_first_stable_index/first_stable_index.cpp
#include <cassert>
#include <tuple>
#include <vector>
using namespace std;

class Solution {
 public:
  int firstStableIndex(vector<int>& nums, int k) {
    int n = nums.size();
    vector<int> minValue(n);
    minValue[n - 1] = nums[n - 1];
    for (int i = n - 2; i >= 0; --i) {
      minValue[i] = min(minValue[i + 1], nums[i]);
    }

    int maxValue = 0;
    for (int i = 0; i < n; ++i) {
      maxValue = max(maxValue, nums[i]);
      if (maxValue - minValue[i] <= k) {
        return i;
      }
    }
    return -1;
  }
};

int main() {
  vector<tuple<vector<int>, int, int>> tests{
      {{5, 0, 1, 4}, 3, 3},
      {{3, 2, 1}, 1, -1},
      {{0}, 0, 0},
  };

  for (auto &[nums, k, ans] : tests) {
    assert(Solution().firstStableIndex(nums, k) == ans);
  }
}

File: golang/3904_first_stable_index/first_stable_index.go
// Package main ...
package main

import "fmt"

func firstStableIndex(nums []int, k int) int {
	n := len(nums)
	if n == 0 {
		return -1
	}

	minValue := make([]int, n)
	minValue[n-1] = nums[n-1]
	for i := n - 2; i >= 0; i-- {
		if minValue[i+1] < nums[i] {
			minValue[i] = minValue[i+1]
		} else {
			minValue[i] = nums[i]
		}
	}

	maxValue := 0
	for i := 0; i < n; i++ {
		if nums[i] > maxValue {
			maxValue = nums[i]
		}
		if maxValue-minValue[i] <= k {
			return i
		}
	}
	return -1
}

func main() {
	tests := []struct {
		nums []int
		k    int
		ans  int
	}{
		{[]int{5, 0, 1, 4}, 3, 3},
		{[]int{3, 2, 1}, 1, -1},
		{[]int{0}, 0, 0},
	}

	for _, test := range tests {
		if got := firstStableIndex(test.nums, test.k); got != test.ans {
			panic(fmt.Sprintf("test %v failed: got %v, want %v", test, got, test.ans))
		}
	}
}
